src/app/services/storage.service.ts

```diff
@@ -386,4 +397,149 @@ export class StorageService {
     }
     return `${bytes.toFixed(dec)} ${units}`;
   }
+
+  getRedundancyLevel(vdev: TopologyItem): number {
+    switch (vdev.type) {
+      case TopologyItemType.Disk:
+        return 0;
+      case TopologyItemType.Mirror:
+        return vdev.children.length - 1;
+      case TopologyItemType.Raidz:
+      case TopologyItemType.Raidz1:
+        return 1;
+      case TopologyItemType.Raidz2:
+        return 2;
+      case TopologyItemType.Raidz3:
+        return 3;
+      default:
+        // VDEV type property also includes values unrelated to layout.
+        return -1;
+    }
+  }
+
+  getVdevWidths(vdevs: TopologyItem[]): Set<number> {
+    const allVdevWidths = new Set<number>(); // There should only be one value
+    vdevs.forEach((vdev) => {
+      let vdevWidthCounter = 0;
+      vdev.children.forEach(() => {
+        vdevWidthCounter += 1;
+      });
+      allVdevWidths.add(vdevWidthCounter);
+    });
+    return allVdevWidths;
+  }
+
+  isMixedWidth(allVdevWidths: Set<number>): boolean {
+    return allVdevWidths.size > 1;
+  }
+
+  // Get usable space on VDEV.
+  getVdevCapacities(vdevs: TopologyItem[]): Set<number> {
+    const allVdevCapacities = new Set<number>(); // There should only be one value
+    vdevs.forEach((vdev) => {
+      allVdevCapacities.add(vdev.stats.size);
+    });
+    return allVdevCapacities;
+  }
+
+  // Check to see if every VDEV has the same capacity. Best practices dictate every vdev should be uniform
+  isMixedVdevCapacity(allVdevCapacities: Set<number>): boolean {
+    return allVdevCapacities.size > 1;
+  }
+
+  getVdevDiskCapacities(vdevs: TopologyItem[], disks: StorageDashboardDisk[]): Set<number>[] {
+    const allDiskCapacities: Set<number>[] = [];
+    vdevs.forEach((vdev) => {
+      const vdevDiskCapacities = new Set<number>(); // There should only be one value
+      if (vdev.children.length) {
+        vdev.children.forEach((child) => {
+          const diskSize = disks?.find((disk) => disk.name === child.disk)?.size;
+          vdevDiskCapacities.add(diskSize);
+        });
+      } else {
+        // Topology items of type DISK will not have children
+        vdevDiskCapacities.add(vdev.stats.size);
+      }
+      allDiskCapacities.push(vdevDiskCapacities);
+    });
+    return allDiskCapacities;
+  }
+
+  // Check to see if any individual VDEVs have non-uniform disk sizes.
+  // Every disk in a VDEV should ideally be the same size
+  isMixedVdevDiskCapacity(allVdevDiskCapacities: Set<number>[]): boolean {
+    const isMixed = allVdevDiskCapacities.filter((vdev) => vdev.size > 1);
+    return isMixed.length > 0;
+  }
+
+  getVdevTypes(vdevs: TopologyItem[]): Set<string> {
+    const vdevTypes = new Set<string>();
+    vdevs.forEach((vdev) => {
+      vdevTypes.add(vdev.type);
+    });
+    return vdevTypes;
+  }
+
+  isMixedVdevType(vdevTypes: Set<string>): boolean {
+    return vdevTypes.size > 1;
+  }
+
+  validateVdevs(category: PoolTopologyCategory,
```

Comment:
I believe you mentioned that there were some scenarios where we may want to compare things between different categories. In this case this method may not work.
You may want to make other methods that are not intended to be used externally `private` and extensively test public methods.

Reply:
In those scenarios we specifically compare to the Data vdevs. Case is covered at line 537 where we check for `TopologyWarning.RedundancyMismatch`. We only do this for Metadata and Dedup VDEVs